docs/coding-guidelines/javascript.md

```diff
@@ -445,574 +504,131 @@ In JavaScript, associative arrays are defined as objects.
-
-    return `${ firstName } ${ lastName }`;
-}
-
-// best
-function getFullName( { firstName, lastName } ) {
-    return `${ firstName } ${ lastName }`;
-}
-```
-
-You can use array destructuring.
-
-```javascript
-const arr = [ 1, 2, 3, 4 ];
-
-// good
-const first = arr[0],
-    second = arr[1];
-
-// better
-const [ first, second ] = arr;
-```
-
-Use object destructuring for multiple return values, not array destructuring.
-
-> Why? You can add new properties over time or change the order of things without breaking call sites.
-
-```javascript
-// bad
-// the caller needs to think about the order of return data
-function processInput( input ) {
-    let left, right, top, bottom;
-    // some assignment happens
-    return [ left, right, top, bottom ];
-}
-const [ left, , top ] = processInput( input );
-
-// good
-// the caller selects only the data they need
-function processInput( input ) {
-    let left, right, top, bottom;
-    // some assignment happens
-    return { left, right, top, bottom };
-}
-const { left, top } = processInput( input );
-```
-
-More about:
-
-- [Destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) at MDN
-
-### Default Parameters
-
-You can use default parameter syntax rather than mutating function arguments.
-
-```javascript
-// really bad
-function handleThings( opts ) {
-    // No! We shouldn't mutate function arguments.
-    // Double bad: if opts is falsy it'll be set to an object which may
-    // be what you want but it can introduce subtle bugs.
-    opts = opts || {};
-    // ...
-}
-
-// good
-function handleThings( opts ) {
-    if ( opts === void 0 ) {
-        opts = {};
-    }
-    // ...
-}
-
-// better
-function handleThings( opts = {} ) {
-    // ...
-}
+```js
+posts.every( ( post ) => post.status === 'publish' );
 ```
 
-More about:
-
-- [Default Parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters) at MDN
-
-### Rest
-
-Never use `arguments`, opt to use rest syntax `...` instead.
-
-> Why? `...` is explicit about which arguments you want pulled. Plus rest arguments are a real Array and not Array-like like `arguments`.
-
-```javascript
-// bad
-function concatenateAll() {
-    const args = Array.prototype.slice.call( arguments );
-
-    return args.join( '' );
-}
+Group an array of post objects by status:
 
-// good
-function concatenateAll( ...args ) {
-    return args.join( '' );
-}
+```js
+posts.reduce( ( memo, post ) => {
+	memo[ post.status ] = ( memo[ post.status ] || [] ).concat( post );
+	return memo;
+}, {} );
 ```
 
-More about:
+### React components
 
-- [Rest](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters) at MDN
+- Use [stateless function components or the `React.Component` class](https://facebook.github.io/react/docs/components-and-props.html#functional-and-class-components) instead of `React.createClass`
```

Comment:
We definitely want to avoid all mixins, and `LinkedStateMixin` has been on the black list for a long time. We have various ways to remove them: higher order components for sure, but also for some things local component state or proper use of our Redux state/query components